seastar/src/config_and_start_seastar.rs

```diff
@@ -0,0 +1,293 @@
+    ///
+    /// assert_eq!(opts.get_description(), description);
+    /// ```
+    pub fn set_description(&mut self, description: &str) {
+        set_description(&self.opts, description);
+    }
+
+    /// Sets the `Options`' number of threads.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use seastar::{AppTemplate, Options};
+    ///
+    /// let mut opts = Options::new();
+    /// let smp = 42;
+    /// opts.set_smp(smp);
+    ///
+    /// assert_eq!(opts.get_smp(), smp);
+    /// ```
+    pub fn set_smp(&mut self, smp: u32) {
+        set_smp(&self.opts, smp);
+    }
+}
+
+impl Default for Options {
+    fn default() -> Self {
+        Options::new()
+    }
+}
+
+/// The object through which the contents of a `main` function would be ran in a Seastar app.
+/// Configurable through [`Options`].
+pub struct AppTemplate {
+    app: UniquePtr<app_template>,
+}
+
+impl AppTemplate {
+    /// Creates a Seastar app based on its configuration (`[seastar-rs::AppTemplate`]).
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use seastar::{AppTemplate, Options};
+    ///
+    /// let app = AppTemplate::new_from_options(Options::default());
+    /// ```
+    pub fn new_from_options(opts: Options) -> Self {
+        AppTemplate {
+            app: new_app_template_from_options(&opts.opts),
+        }
+    }
+
+    /// Runs an app with a void callback (the output of which is always 0) and program arguments (argv).
+    /// The app is run on a separate thread.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use seastar::{AppTemplate, Options};
+    ///
+    /// let func = || { println!("{}", 42); };
+    ///
+    /// let app = AppTemplate::default();
+    /// let args = vec!["hello".to_owned()];
+    ///
+    /// assert_eq!(app.run_void(&args, func), 0);
+    /// ```
+    pub fn run_void(&self, args: &Vec<String>, func: fn()) -> i32 {
+        run_void(&self.app, args, func)
+    }
+
+    /// Runs an app with an int (status code) callback and program arguments (argv).
+    /// The app is run on a separate thread.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use seastar::{AppTemplate, Options};
+    ///
+    /// let func = || { 42 };
+    ///
+    /// let app = AppTemplate::default();
+    /// let args = vec!["hello".to_owned()];
+    ///
+    /// assert_eq!(app.run_int(&args, func), 42);
+    /// ```
+    pub fn run_int(&self, args: &Vec<String>, func: fn() -> i32) -> i32 {
+        run_int(&self.app, args, func)
+    }
+}
+
+impl Default for AppTemplate {
+    fn default() -> Self {
+        AppTemplate::new_from_options(Options::default())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_new_options_contain_default_values() {
+        let opts = Options::new();
+        assert_eq!(opts.get_name(), "App");
+        assert_eq!(opts.get_description(), "");
+        assert_eq!(opts.get_smp(), num_cpus::get() as u32);
```

Comment:
Are `opts.get_smp()` and `num_cpus::get()` guaranteed to return the same on all systems? Do they return exactly the same thing?

Reply:
Seastar:
```cpp
/// Configuration for the multicore aspect of seastar.
struct smp_options : public program_options::option_group {
    /// Number of threads (default: one per CPU).
    program_options::value<unsigned> smp;
 ...
 ```
 `num_cpus::get`: "This function will get the number of logical cores. [...]"
Which I assumed (correctly?) to mean "threads".